### PR TITLE
nixos/shadow: introduce security.shadow.enable

### PR DESCRIPTION
**Pull request by tomfitzhenry**
Allow users to disable the shadow authentication suite.

My primary motivation is to reduce the attack surface of setuid binaries, by removing them where possible. shadow understandably introduces many setuid binaries. I realised, however, that I don't use any of these.

The test demonstrates login working without needing the shadow suite.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Thesola10** [comment]
I'd add an assertion for `shadow.enable=false -> greetd.enable = true`

File: nixos/modules/programs/shadow.nix [hunk 1]
```diff
@@ -138,7 +139,7 @@ in
 
   ###### implementation
 
-  config = {
+  config = lib.mkIf config.security.shadow.enable {
```
Comment (Thesola10) [inline nixos/modules/programs/shadow.nix:142]:
```suggestion
  config = 
   let myAssert = a: if config.security.shadow.enable == false -> config.services.greetd.enable then a else throw "One of config.security.shadow or config.services.greetd must be set.";
   in
   myAssert lib.mkIf config.security.shadow.enable {
```


This is how I'd do it.

File: nixos/modules/programs/shadow.nix [hunk 2]
```diff
@@ -5,6 +5,7 @@ let
 in
 {
   options = with lib.types; {
+    security.shadow.enable = lib.mkEnableOption "the shadow authentication suite" // { default = true; };
```
Comment (SuperSandro2000) [inline nixos/modules/programs/shadow.nix:8]: This needs an explanation that it is experimental and things might not work correctly if enabled.

Also could this lock yourself out of your system?

File: nixos/modules/programs/shadow.nix (hunk 1 above)
Comment (SuperSandro2000) [reply nixos/modules/programs/shadow.nix:142]: We don't want to use throw in the module system. This should be done in assertions.

**Comment by SuperSandro2000** [comment]
> SGTM. I looked into doing that. I'm familiar with `assertions`, but the difficulty is making sure that the assertion runs, even if `config.security.shadow.enable` is false.

You are searching for lib.mkMerge

File: nixos/modules/programs/shadow.nix [hunk 3]
```diff
@@ -5,6 +5,18 @@ let
 in
 {
   options = with lib.types; {
+
+    security.shadow.enable = lib.mkOption {
```
Comment (SuperSandro2000) [inline nixos/modules/programs/shadow.nix:9]: We should probably use mkEnableOption // { description = ''.....''; } instead

File: nixos/modules/programs/shadow.nix [hunk 4]
```diff
@@ -5,6 +5,18 @@ let
 in
 {
   options = with lib.types; {
```
Comment (SuperSandro2000) [inline nixos/modules/programs/shadow.nix:7]:
```suggestion
  options = {
```
While we are at it, we should drop this and just write lib.types.

File: nixos/modules/programs/shadow.nix [hunk 5]
```diff
@@ -138,101 +150,111 @@ in
...
-        "login.defs".source = pkgs.writeText "login.defs" (toKeyValue cfg.settings);
-
-        # /etc/default/useradd: configuration for useradd.
-        "default/useradd".source = pkgs.writeText "useradd" ''
+  config = lib.mkMerge [
+    {
+      assertions = [
+        {
+          assertion = config.security.shadow.enable || config.services.greetd.enable;
+          message = "You must enable a VT login method: security.shadow.enable or services.greetd.enable";
```
Comment (SuperSandro2000) [inline nixos/modules/programs/shadow.nix:158]:
```suggestion
          message = "You must enable at least one VT login method, either security.shadow.enable or services.greetd.enable";
```

File: nixos/modules/programs/shadow.nix [hunk 6]
```diff
@@ -138,101 +150,111 @@ in
...
+          message = "GID_MIN must be less than or equal to GID_MAX";
+        }
+      ];
+
+      security.loginDefs.settings.CHFN_RESTRICT =
+        lib.mkIf (cfg.chfnRestrict != null) cfg.chfnRestrict;
+
+      environment.systemPackages = lib.optional config.users.mutableUsers cfg.package
+                                   ++ lib.optional (lib.types.shellPackage.check config.users.defaultUserShell) config.users.defaultUserShell
+                                   ++ lib.optional (cfg.chfnRestrict != null) pkgs.util-linux;
```
Comment (SuperSandro2000) [inline nixos/modules/programs/shadow.nix:187]:
```suggestion
      security.loginDefs.settings.CHFN_RESTRICT = lib.mkIf (cfg.chfnRestrict != null) cfg.chfnRestrict;

      environment.systemPackages = lib.optional config.users.mutableUsers cfg.package
        ++ lib.optional (lib.types.shellPackage.check config.users.defaultUserShell) config.users.defaultUserShell
        ++ lib.optional (cfg.chfnRestrict != null) pkgs.util-linux;
```

File: nixos/modules/programs/shadow.nix [hunk 7]
```diff
@@ -138,101 +150,111 @@ in
...
+            # /etc/login.defs: global configuration for pwdutils.
+            # You cannot login without it!
+            "login.defs".source = pkgs.writeText "login.defs" (toKeyValue cfg.settings);
+
+            # /etc/default/useradd: configuration for useradd.
+            "default/useradd".source = pkgs.writeText "useradd" ''
           GROUP=100
           HOME=/home
           SHELL=${utils.toShellPath config.users.defaultUserShell}
         '';
```
Comment (SuperSandro2000) [inline nixos/modules/programs/shadow.nix:207]:
```suggestion
            "default/useradd".source = pkgs.writeText "useradd" ''
              GROUP=100
              HOME=/home
              SHELL=${utils.toShellPath config.users.defaultUserShell}
            '';
```

File: nixos/modules/programs/shadow.nix [hunk 8]
```diff
@@ -138,101 +150,111 @@ in
...
+        groupmod.rootOK = true;
+        groupmems.rootOK = true;
+        groupdel.rootOK = true;
+        login = {
+          startSession = true;
+          allowNullPassword = true;
+          showMotd = true;
+          updateWtmp = true;
+        };
+        chpasswd = { rootOK = true; };
```
Comment (SuperSandro2000) [inline nixos/modules/programs/shadow.nix:235]:
```suggestion
        chpasswd.rootOK = true;
```

File: nixos/modules/programs/shadow.nix [hunk 9]
```diff
@@ -138,101 +150,111 @@ in
...
-    security.wrappers =
-      let
-        mkSetuidRoot = source: {
-          setuid = true;
-          owner = "root";
-          group = "root";
-          inherit source;
+      security.pam.services = {
+        chsh = { rootOK = true; };
+        chfn = { rootOK = true; };
```
Comment (SuperSandro2000) [inline nixos/modules/programs/shadow.nix:212]:
```suggestion
        chsh.rootOK = true;
        chfn.rootOK = true;
```

**Review by SuperSandro2000 — APPROVED** [review]
LGTM

**Pull request merged**